### ISSUE TITLE
Improve guarding of Bytes and other classes

### ISSUE THREAD

**Issue by minborg** [OpenHFT/Chronicle-Core#273]
Would it be better if we had a pattern like

```
try (Reference ref = bytes.acquireReferenceOrThrowIfReleased()) {
   ... do stuff
}
```

instead of `throwExceptionIfReleased`, that way we'd be protected against the thing being released while we're in the block as well? 

It would also have the behaviour that if you were passed a released `Bytes` it would throw.

**Comment by peter-lawrey** [comment]
It could be done object free as we have two mode, with object tracing we create some garbage and have higher overhead checks, however without this option we have a less strict but still lightweight check.

**Comment by peter-lawrey** [comment]
The current pattern is 

```
        ReferenceOwner toDebugString = ReferenceOwner.temporary("toDebugString");
        bytes.reserve(toDebugString);
        try {
            /// use bytes
        } finally {
            bytes.release(toDebugString);
        }
```
When tracing is on, this creates an object each time and does full checking of matching reserve/release.
When tracing is off, the object is actually `null` and the methods just increment and decrement the `refCount`

**Comment by minborg** [comment]
Maybe the current pattern is good enough.

**Comment by minborg** [comment]
Ideally, the unguarded methods could be moved to the acquired Reservation so they only can be accessed when we have acquired a Reservation (which is safe) and not directly (which is unsafe)

**Comment by minborg** [comment]
This could be input to a general rewrite of the memory abstraction layer. I am closing this as that is a separate issue.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Core-2.27ea8](https://github.com/OpenHFT/Chronicle-Core/releases/chronicle-core-2.27ea8), [BOM-2.27ea82](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.27ea82)
